task/udprpc/src/main.rs

```diff
@@ -0,0 +1,147 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#![no_std]
+#![no_main]
+
+use task_net_api::*;
+use userlib::*;
+use zerocopy::{AsBytes, FromBytes};
+
+task_slot!(NET, net);
+
+#[derive(Copy, Clone, Debug)]
+#[repr(u8)]
+enum RpcReply {
+    Ok,
+    /// The RPC packet was too short to include the complete header
+    TooShort,
+    /// The RPC packet's image ID does not match ours
+    BadImageId,
+    /// The size of the packet does not agree with the number of bytes specified
+    /// in the `nbytes` field of the packet
+    NBytesMismatch,
+    /// The `nbytes` field of the packet imply that the packet overflows the
+    /// `rx_data_buf`
+    NBytesOverflow,
+    /// The result of the function does not agree with the number of bytes
+    /// specified in the `nreply` field of the packet
+    NReplyMismatch,
+    /// The output would overflow `tx_data_buf`
+    NReplyOverflow,
+}
+
+/// Header for an RPC request
+///
+/// `humility` must cooperate with this layout, which is mirrored in `doppel.rs`
+#[derive(Copy, Clone, Debug, FromBytes)]
+#[repr(C)]
+struct RpcHeader {
+    image_id: u64,
+    task: u16,
+    op: u16,
+    nreply: u16,
+    nbytes: u16,
+}
+
+#[export_name = "main"]
+fn main() -> ! {
+    let net = NET.get_task_id();
+    let net = Net::from(net);
+
+    const SOCKET: SocketName = SocketName::rpc;
+    let image_id = kipc::read_image_id();
+
+    // We use the image id to make sure that we're compatible.
+    //
+    // The output format is dependent on status code.  The first byte is always
+    // a member of `RpcReply` as a `u8`.
+    // - `NBytesMismatch`, `NBytesOverflow`, `NReplyMismatch`, `NReplyOverflow`
+    //   return nothing else (so the reply is 1 byte)
+    // - `BadImageId` is followed by the *actual* 64-bit image id as a big-endian
+    //   value
+    // - `Ok` is followed by the return code as a 32-bit, big-endian value,
+    //   then by `nreply` bytes of reply.
+    loop {
+        let mut rx_data_buf = [0u8; 1024];
+        let mut tx_data_buf = [0u8; 1024];
+        match net.recv_packet(
+            SOCKET,
+            LargePayloadBehavior::Discard,
+            &mut rx_data_buf,
+        ) {
+            Ok(mut meta) => {
+                // We can always read the header, because it's raw data.  It may
+                // not be valid, though, if the packet is too short (checked
+                // in the conditional below).
+                const HEADER_SIZE: usize = core::mem::size_of::<RpcHeader>();
+                let header =
+                    RpcHeader::read_from(&rx_data_buf[..HEADER_SIZE]).unwrap();
+
+                const REPLY_PREFIX_SIZE: usize = 5;
+
+                // We deliberately assign to `r` here then manipulate it;
+                // otherwise, the compiler won't include RpcReply in DWARF data.
+                let r = if (meta.size as usize) < HEADER_SIZE {
```

Comment:
I was going to leave a comment about not checking `meta.size` before constructing `header`, but I see it here; it's slightly strange to me for it to be in this order. Maybe throw a comment on the `read_from` call above that we'll check the size below (before reading any fields of the potentially-invalid `header`)?

Reply:
I ended up rearranging this a little bit to make it less weird!